lib/algolia.ex

```diff
@@ -491,6 +491,149 @@ defmodule Algolia do
     |> inject_index_into_response(index)
   end
 
+  @doc """
+  Get all the synonyms of a index.
+
+  Only the index is required
+
+  ## Examples
+
+      iex> Algolia.export_synonyms("index")
+      {:ok, %{
+        "indexName" => "index",
+        "nbHits" => 33,
+        "hits" => [%{
+          "objectID" => "1539816124475_0",
+          "synonyms" => ["big", "large"],
+          "type" => "synonym"
+          "_highlightResult" => %{...}
+        }, ...]
+      }}
+  """
+  def export_synonyms(index) do
```

Comment:
The official clients seem to have a iterative approach to these functions (like the Go client and the Scala client), which allows us to process one hit at a time. The Ruby one goes a step further and also returns the complete list at the end. I think we can achieve a similar flexibility in Elixir by leveraging Streams.

Another detail that I noticed in those clients is that they drop the `_highlightResult` key from every hit.

So I came up with this sketch of what could be a stream-based implementation:

```elixir
def export_synonyms(index) do
  get_all_paginated_hits(index, &search_synonyms/2)
end

def get_all_paginated_hits(index, search, hits_per_page \\ 100) do
  0
  |> Stream.iterate(&(&1 + 1))
  |> Stream.flat_map(fn page -> get_page_hits(index, search, page, hits_per_page) end)
  |> Stream.take_while(&(&1 != :stop))
end

defp get_page_hits(index, search, page, hits_per_page) do
  case search.(index, "", page: page, hits_per_page: hits_per_page) do
    {:ok, %{"hits" => hits, "nbPages" => pages}} when page + 1 < pages -> page_hits(hits)
    {:ok, %{"hits" => hits}} -> page_hits(hits) ++ [:stop]
    error -> [error, :stop]
  end
end

defp page_hits(hits) do
  Enum.map(hits, &{:ok, Map.drop(&1, ["_highlightResult"])})
end
```

This would return a `Stream` and thus allow users to consume it as they wish, including error handling. For example:

```elixir
"my_index"
|> Algolia.export_synonyms()
|> Enum.each(fn
  {:ok, %{"synonyms" => [from, to]}} -> IO.puts("#{from} and #{to} are the same")
  {:error, _} -> IO.puts("Unable to fetch all synonyms.")
end)
```

Or they can go ahead and just get the whole list, raising a match error if one of the page requests fails:

```elixir
"my_index"
|> Algolia.export_synonyms()
|> Enum.map(fn -> {:ok, hit} -> hit end)
```

How do you feel about that?